nx/lib/nx/pytorch_backend.ex

```diff
@@ -0,0 +1,177 @@
+defmodule Nx.PytorchBackend do
+  @behaviour Nx.Backend
+
+  defstruct [:ref]
+
+  alias Nx.Pytorch.NIF
+
+  import Nx.Shared
+
+  funs =
+    Nx.Backend.behaviour_info(:callbacks) --
+      [
+        tensor: 1,
+        iota: 2,
+        eye: 1,
+        random_uniform: 3,
+        random_normal: 3,
+        reshape: 3,
+        squeeze: 3,
+        add: 3,
+        dot: 5,
+        cholesky: 2,
+        from_binary: 3,
+        to_binary: 2,
+        backend_deallocate: 1,
+        inspect: 2
+      ]
+
+  for {fun, arity} <- funs do
+    args = Macro.generate_arguments(arity, __MODULE__)
+
+    def unquote(fun)(unquote_splicing(args)) do
+      raise "#{unquote(fun)}() is not supported"
+    end
+  end
+
+  ## Creation
+
+  @impl true
+  def tensor(tensor) do
+    tensor
+  end
+
+  def torch_type({:u, 8}), do: :byte
+  def torch_type({:s, 8}), do: :char
+  def torch_type({:s, 16}), do: :short
+  def torch_type({:s, 32}), do: :int
+  def torch_type({:s, 64}), do: :long
+  def torch_type({:f, 16}), do: :half
+  def torch_type({:f, 32}), do: :float
+  def torch_type({:f, 64}), do: :double
+
+  @impl true
+  def iota(out, axis \\ nil)
+
+  def iota(%{shape: {}, type: type} = out, nil), do: iota(%{out | shape: {1}}, nil)
+
+  def iota(%{shape: shape, type: type} = out, nil) do
+    t = NIF.arange(0, Nx.size(shape), 1, torch_type(type), shape)
+    from_ref(out, t)
+  end
+
+  def iota(%{shape: {n}, type: type} = out, 0) do
+    t = NIF.arange(0, n, 1, torch_type(type))
+    from_ref(out, t)
+  end
+
+  def iota(%{shape: shape, type: type} = out, axis) do
+    {dims_before, [dim | dims_after]} =
+      shape
+      |> Tuple.to_list()
+      |> Enum.split(axis)
+
+    # Number of repetitions of an index in memory
+    repeat_blocks =
+      dims_after
+      |> Enum.reduce(1, &*/2)
+
+    # Number of cycles of the counting pattern
+    cycles =
+      dims_before
+      |> Enum.reduce(1, &*/2)
+
+    data =
+      for _ <- 1..cycles,
+          i <- 0..(dim - 1),
+          _ <- 1..repeat_blocks,
+          into: "",
+          do: number_to_binary(i, type)
+
+    t = NIF.from_blob(data, shape, torch_type(type))
+    from_ref(out, t)
```

Comment:
So there is a faster implementation on top of aten which is by doing this. Imagine you have a tensor `t`, with `shape`, and you want iota on `axis`. You can do this:

```elixir
# gets the size of iota
dim = elem(shape, axis)

# build the iota in one dimension
aten = NIF.arange(0, dim)

# reshape the tensor above to be have shape where everything is 1, except for dim
reshape = Tuple.duplicate(1, Nx.rank(shape)) |> put_elem(axis, dim)
aten = NIF.reshape(aten, reshape)

# Now broadcast the tensor using the original shape
aten = NIF.broadcast(aten, shape)
```

You can see this working with Nx too:

```elixir
iex> Nx.iota({3}) |> Nx.reshape({1, 1, 3}) |> Nx.broadcast({3, 3, 3}) == Nx.iota({3, 3, 3}, axis: 2)
true

iex> Nx.iota({3}) |> Nx.reshape({3, 1, 1}) |> Nx.broadcast({3, 3, 3}) == Nx.iota({3, 3, 3}, axis: 0)
true
```

This will most likely be more efficient on aten side, especially if it lazily expands broadcasts. :)